### PR TITLE
feat: notify cancel

### PR DESCRIPTION
**Pull request by wischli**
### Product requirements

* Before this change, `cancel{Deposit, Redeem}` is the only action triggered from a Spoke-to-Hub message that can also initiate a message to be sent back a Hub-to-Spoke message. The message would be prepared as `unpaid` and requires manual payment. To improve UX, @lemunozm proposed to change the behavior from manually requiring repayment to manually creating and sending the `Fulfilled*` message with the cancellation amount ([thread](https://kflabs.slack.com/archives/C07PG2EUR9C/p1758895041376429?thread_ts=1758793226.119319&cid=C07PG2EUR9C))

### Design notes

* Adds `notifyCancel{Deposit, Redeem}` messages to `IHubRequestManagerNotifications` and `BatchRequestManager`
  * Permissionless but `protected` similar to `notify{Deposit, Redeem}`
  * Only handles immediate cancellations, queued cancellation flow is unchanged and still needs to be resolved via `notify{Deposit, Redeem}`
  * Required adding new storage -> Alternative would be modifying `UserOrder` struct to include immediate pending cancellation amounts

### TODOs

- [ ] Add missing coverage for edge cases -> can be done in follow-up PR

### CODE REVIEW

File: test/integration/EndToEnd.t.sol [hunk 1]
```diff
@@ -374,6 +374,11 @@ contract EndToEndUtils is EndToEndDeployment {
     function _getLastUnpaidMessage() internal returns (bytes memory message) {
         Vm.Log[] memory logs = vm.getRecordedLogs();
 
+        if (logs.length == 0) {
+            vm.recordLogs();
+            return message;
+        }
```
Comment (lemunozm) [inline test/integration/EndToEnd.t.sol]: I don't understand this. Is it some test remnant?

File: test/integration/EndToEnd.t.sol [hunk 2]
```diff
@@ -870,8 +875,12 @@ contract EndToEndFlows is EndToEndUtils {
         vault.requestRedeem(shares, INVESTOR_A, INVESTOR_A);
         vault.cancelRedeemRequest(PLACEHOLDER_REQUEST_ID, INVESTOR_A);
 
-        if (!sameChain) h.gateway.repay{value: GAS}(s.centrifugeId, _getLastUnpaidMessage());
+        vm.startPrank(ANY);
+        vm.deal(ANY, GAS);
```
Comment (lemunozm) [inline test/integration/EndToEnd.t.sol]: NIT. `ANY` already has 1 ether

File: test/integration/EndToEnd.t.sol [hunk 3]
```diff
@@ -1111,11 +1120,14 @@ contract EndToEndUseCases is EndToEndFlows, VMLabeling {
         vault.requestDeposit(USDC_AMOUNT_1, INVESTOR_A, INVESTOR_A);
         vault.cancelDepositRequest(PLACEHOLDER_REQUEST_ID, INVESTOR_A);
 
-        if (!sameChain) h.gateway.repay{value: GAS}(s.centrifugeId, _getLastUnpaidMessage());
+        vm.startPrank(ANY);
+        vm.deal(ANY, GAS);
```
Comment (lemunozm) [inline test/integration/EndToEnd.t.sol]: Also here

**Review by lemunozm — DISMISSED** [review]
LGTM, as long as this https://kflabs.slack.com/archives/C07PG2EUR9C/p1759162992576159?thread_ts=1758793226.119319&cid=C07PG2EUR9C works correctly.

Thanks for taking this! I can take the NITs in my PR if you agree

File: test/integration/EndToEnd.t.sol (hunk 1 above)
Comment (wischli) [reply test/integration/EndToEnd.t.sol]: Ah yes sorry, this was part of debugging. Removed with the `vm.deal`s in [d85c43e](https://github.com/centrifuge/protocol-v3/pull/640/commits/d85c43e7db2851dadf05bd5e1e18efb9b45a464b)

**Comment by wischli** [comment]
Closing due to insufficient UX improvement
